cmd/thanos/rule.go

```diff
@@ -379,40 +377,19 @@ func runRule(
 		})
 	}
 
-	var (
-		bkt    objstore.Bucket
-		bucket string
-		// closeFn gets called when the sync loop ends to close clients, clean up, etc
-		closeFn = func() error { return nil }
-		uploads = true
-	)
+	var uploads bool = true
 
 	// The background shipper continuously scans the data directory and uploads
 	// new blocks to Google Cloud Storage or an S3-compatible storage service.
-	if gcsBucket != "" {
-		gcsClient, err := storage.NewClient(context.Background())
-		if err != nil {
-			return errors.Wrap(err, "create GCS client")
-		}
-
-		bkt = gcs.NewBucket(gcsBucket, gcsClient.Bucket(gcsBucket), reg)
-		closeFn = gcsClient.Close
-		bucket = gcsBucket
-	} else if s3Config.Validate() == nil {
-		bkt, err = s3.NewBucket(s3Config, reg, component)
-		if err != nil {
-			return errors.Wrap(err, "create s3 client")
-		}
-
-		bucket = s3Config.Bucket
+	bkt, closeFn, err := client.NewBucket(&gcsBucket, *s3Config, reg, component)
+	if err != nil && err != client.ErrNotFound {
+		return err
 	} else {
```

Comment:
this is ambigous. I think flow will enter it if err == nil

Can we explicitly do `if err == client.ErrNotFound` here?

Reply:
like
```go
	if err == client.ErrNotFound {
		level.Info(logger).Log("msg", "No GCS or S3 bucket configured, uploads will be disabled")
		uploads = false
	} else if err != nil {
		return err
	}
```
I think otherwise we wouldn't return on error

Reply:
Not really. Whatever that works and use least possible amount of `else` as possible (: (why else if you out-ed some case because your did `return err`?

For example:
```
if err != nil && err != client.ErrNotFound {
  return err
}

if err == client.ErrNotFound {
  // do some stuff
}

```